java-checks/src/main/java/org/sonar/java/checks/SynchronizationOnStringOrBoxedCheck.java

```diff
@@ -52,17 +89,29 @@
   @Override
   public void visitNode(Tree tree) {
     SynchronizedStatementTree syncStatement = (SynchronizedStatementTree) tree;
-    Type expressionType = syncStatement.expression().symbolType();
-    if (expressionType.isPrimitive() || isForbiddenType(expressionType)) {
-      reportIssue(syncStatement.expression(), "Synchronize on a new \"Object\" instead.");
+    ExpressionTree expression = syncStatement.expression();
+    Type expressionType = expression.symbolType();
+    if (expressionType.isPrimitive() || isForbiddenType(expressionType) || isInitializedWithImmutable(expression)) {
+      reportIssue(expression, "Synchronize on a new \"Object\" instead.");
     }
   }
 
   private static boolean isForbiddenType(Type expressionType) {
-    for (String forbiddenType : FORBIDDEN_TYPES) {
-      if (expressionType.is(forbiddenType)) {
-        return true;
-      }
+    return FORBIDDEN_TYPES.contains(expressionType.fullyQualifiedName()) 
+      || FORBIDDEN_SUBTYPES.stream().anyMatch(expressionType::isSubtypeOf);
+  }
+  
+  private static boolean isInitializedWithImmutable(ExpressionTree tree) {
```

Comment:
The idea is to check that the expression used in <code>synchronized()</code> is either call to <code>listOf() setOf()...</code> or variable/field initialized with these methods.

As far as copmpile-time type is <code>List</code> it's impossible to check wether it's <code>ListN</code>, <code>SetN</code> or <code>MapN</code> in Runtime.